### PR TITLE
Switch to Graph data structure

### PR DESCRIPTION
**Pull request by waqarsaleem**
Rewrote the entire program
- to use builtin `networks.DiGraph` instead of custom data type
- incorporated visualization of preference graph
- renamed file to avoid conflict with python's builtin `select` module
- used some newer python features: assignment statement, type annotation, *-unpacking

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,40 +1,55 @@
...
+I need to make student teams of 2 in my classes for which I ask students to specify _greens_ and _reds_. Greens are people that the student wants to work with and reds are people that they _do not_ want to work with. This program matches students according to these prefernces.
 
 ## How to Use
 
-You _only_ need to make the following edits in the `init_students` function toward the bottom of the file,  _select.py_.
-1. Add the names of _all_ the students in your class who need to be paired to the indicated list. This includes students who did not indicate any preference, e.g. because they did not fill the required form by the deadline, but still need to be paired. The names _must be unique_.
-1. Enter the preferences using the `add_green` and `add_red` functions. Their syntax is as follows.
+You _only_ need to make the filloiwng edits in the `enter_data` function toward the bottom of the file,  _match.py_. The segments to be edited are clearly marked in the file with __EDIT__.
```
Comment (ma569) [inline README.md]: minor: following*

minor: may be you don't need this"...toward the bottom of the file"

File: README.md [hunk 2]
```diff
@@ -1,40 +1,55 @@
...
 
-The program tries to partition the vertices into pairs such that no pair contains an enemy. It does so by randomly selecting an unpaired vertex as _source_ and selecting as _destination_ a friend for whom _source_ is not an enemy. If _destination_ is already part of another pair, a new _destination_ is selected. If no such selection is possible, _destination_ is selected from the unpaired vertices that are not directly connected to _source_. Once an acceptable _destination_ is found, the pair (_source_,  _destination_) is added to the existing set of pairs. The set of pairs is called a _pairing_. A pairing is scored in terms of the number of friends that are paired in it. A perfect pairing is one in which all pairs consist of friends.
+The program builds a matching by iteratively picking a random node as the _source_ and buildong a match with one of its greens such that _source_ is not a red of the green. If a match cannot be found among the greens, nodes not connected to _source_ are searched for matches. If still no match is found, the matching fails. Once a match is found, _source_ and its match are added to the matching and removed from the graph.
```
Comment (ma569) [inline README.md]: minor: building not buildong

File: README.md [hunk 3]
```diff
@@ -1,40 +1,55 @@
...
 
-The above process is repeated several times to maximize the likelihood of a high scoring pairing. The highest scoring pairings (there may be multiple) are saved.
+AS random choices are involved, the above process is repeated several times to maximize the likelihood of a high scoring pairing. The highest scoring pairings (there may be multiple) are saved.
 
 ## Limitations
 
-I believe that it is possible, theoretically, for the initial set of preferences to not lead to a solution, i.e. it is impossible to pick a pair that does not include an enemy. However, I have not encountered that in practice.
-
-I have also not considered how this works for an odd number of students. The program could be run for an even number of students, leaving one student out, and that remaining student could be inserted manually into a computed pair.
+- It may be possible for a set of prefernces to not lead to any possible matching. That has to be studied.
```
Comment (ma569) [inline README.md]: minor: preferences

File: match.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            dot.edge(src,dst, color='darkgreen' )
+        else:
+            dot.edge(src,dst, color='red' )
+    dot.format = 'png'
+    dot.render(file_name, view=True)
+            
+def enter_data(G : nx.DiGraph) -> None:
+    '''Populates the digraph with student preferences.'''
+
+    def add_targets(src : str, greens : [str], reds : [str]):
```
Comment (ma569) [inline match.py]: I don't understand why use a nested function here? Might as well make it global for easier readability. Particularly, when you are asking users to edit the code file in the right next to it

File: match.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    ### EDIT: Add greens and reds below.
+    add_targets("U", ["Z", "P", "W"], ["T"])
+    add_targets("X", ["Z", "P", "Q"], ["T"])
+    add_targets("Q", ["X", "Z", "R"], ["T"])
+    add_targets("R", ["V", "S", "U"], ["T"])
+    add_targets("S", ["V", "P", "U"], ["T"])
+    add_targets("W", ["U", "P", "S"], ["T"])
+    add_targets("Z", ["S", "U", "P"], ["T"])
+    add_targets("V", ["R", "U", "P"], ["W"])
+    # add_targets("Y", ["P", "R", "V"], ["T"])
```
Comment (ma569) [inline match.py]: Ideally, you either take input from the command line. Or you define your input structure at the top of the file. User should not be entering input in the middle of the code. Hence, my earlier suggestion that make user enter the data at the top of the file and you can call addTarget function. 

I am sure you have a reason for this which I am mission

File: match.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    dot.render(file_name, view=True)
+            
+def enter_data(G : nx.DiGraph) -> None:
+    '''Populates the digraph with student preferences.'''
+
+    def add_targets(src : str, greens : [str], reds : [str]):
+        for n in greens:
+            G.add_edge(src, n, pref=True)
+        for n in reds:
+            G.add_edge(src, n, pref=False)
```
Comment (ma569) [inline match.py]: Question: Why add false explicitly? Not adding an edge amounts to automatic red?

File: match.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    from graphviz import Digraph
+    file_name = "preferences"
+    dot = Digraph()
+    dot.attr('edge', color = 'darkgreen:red')
+    for src,dst in G.edges():
+        if G[src][dst]['pref']:
+            dot.edge(src,dst, color='darkgreen' )
+        else:
+            dot.edge(src,dst, color='red' )
+    dot.format = 'png'
```
Comment (ma569) [inline match.py]: yikes png. Should be vector graphics :)

File: match.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    G.add_nodes_from(no_data)
+    
+### WORKS WELL ON ITS OWN. EDIT IF YOU KNOW WHAT YOU ARE DOING. 
+def main():
+    preferences = nx.DiGraph()
+    enter_data(preferences)
+    visualize(preferences)
+    for n in preferences.nodes():
+        preferences.nodes[n].setdefault('pref', False)
+    num_tries : int = 10000
```
Comment (ma569) [inline match.py]: these magic numbers must be at the top of the file or in main

File: match.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    high_score : int = -1
```
Comment (ma569) [inline match.py]: these magic numbers must be at the top of the file or in main. 
instead of -1, you could use python min

File: match.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+def startProgress(title : str):
+    global progress_x
+    sys.stdout.write(title + ": [")
+    sys.stdout.flush()
+    progress_x = 0
+
+def progress(x : int):
+    global progress_x
+    x = int(x * 40 // 100)
```
Comment (ma569) [inline match.py]: magic numbers

**Review by ma569 — CHANGES_REQUESTED** [review]
I would suggest that we should add some test cases.

I am not sure how you have implemented 0.6 score in the code

File: match.py (hunk 4 above)
Comment (waqarsaleem) [reply match.py]: The nested function is this function's helper. No other function needs it so nesting it avoids unnecessary clutter of the global namespace. Good point about user accidentally breaking it as they have to edit neighboring lines. That is now gone. The program reads from csv. The user no longer needs to fiddle with the code.

File: match.py (hunk 5 above)
Comment (waqarsaleem) [reply match.py]: Updated to read from a CSV file.

File: match.py (hunk 6 above)
Comment (waqarsaleem) [reply match.py]: A missing edge is different from a red edge. More specifically, a missing edge is a candidate  for matching if all green edges fail, but a red edge is not.

File: match.py (hunk 7 above)
Comment (waqarsaleem) [reply match.py]: Updated to pdf, which is the default.

File: match.py (hunk 9 above)
Comment (waqarsaleem) [reply match.py]: Python 3 _tries_ to model[ unbounded integers so the concept of the smallest integer is hairy][1]. However I have used a reasonable suggestion from the thread which should be fine as all the numbers that are compared with are positive.

[1]: https://stackoverflow.com/questions/7604966/maximum-and-minimum-values-for-ints

File: README.md (hunk 1 above)
Comment (waqarsaleem) [reply README.md]: Did a spell check and rewrote the section with instructions for the CSV file.

File: README.md (hunk 2 above)
Comment (waqarsaleem) [reply README.md]: Spell checked. 0.6 is not incorporated, rather indicated as a guide for the user to make sense of the output. The output contains a _score_ between 0 and 1. The number may not make much sense to the user so 0.6 is proposed as a good score.

File: README.md (hunk 3 above)
Comment (waqarsaleem) [reply README.md]: Spell checked.

File: match.py (hunk 8 above)
Comment (waqarsaleem) [reply match.py]: Moved to the top.

File: match.py (hunk 10 above)
Comment (waqarsaleem) [reply match.py]: 100 is for percent. These are for progress bar logic which is not very important for now. Letting it be for now.

File: README.md (hunk 2 above)
Comment (waqarsaleem) [reply README.md]: Regarding test cases, a sample CSV is provided and the README is updated to contain the corresponding output and its explanation. A note on the visualization is also included.

**Comment by waqarsaleem** [comment]
Catered to suggestions. Thank you for them. I imagined this step to be a formality but I am glad to have received detailed feedback. Thanks again for your efforts. They led to better code.

**Review by ma569 — APPROVED** [review]

**Pull request merged**